src/Symfony/Component/Process/ProcessUtils.php

```diff
@@ -32,46 +34,42 @@ private function __construct()
     /**
      * Escapes a string to be used as a shell argument.
      *
+     * Provides a more robust method on Windows than escapeshellarg.
+     * See https://blogs.msdn.microsoft.com/twistylittlepassagesallalike/2011/04/23/everyone-quotes-command-line-arguments-the-wrong-way/
+     *
      * @param string $argument The argument that will be escaped
+     * @param int    $mode     A bitfield of self::ESC_WINDOWS_* constants to configure escaping context on Windows
      *
      * @return string The escaped argument
      */
-    public static function escapeArgument($argument)
+    public static function escapeArgument($argument, $mode = self::ESC_WINDOWS_ARGV_CMD)
     {
-        //Fix for PHP bug #43784 escapeshellarg removes % from given string
-        //Fix for PHP bug #49446 escapeshellarg doesn't work on Windows
-        //@see https://bugs.php.net/bug.php?id=43784
-        //@see https://bugs.php.net/bug.php?id=49446
-        if ('\\' === DIRECTORY_SEPARATOR) {
-            if ('' === $argument) {
-                return escapeshellarg($argument);
-            }
+        if ('\\' !== DIRECTORY_SEPARATOR) {
+            return escapeshellarg($argument);
+        }
+        if (!(self::ESC_WINDOWS_ARGV_CMD & $mode)) {
+            throw new \InvalidArgumentException(sprintf('The $mode argument of %s must be a non-zero bitfield of self::ESC_WINDOWS_* constants.', __METHOD__));
+        }
 
-            $escapedArgument = '';
-            $quote = false;
-            foreach (preg_split('/(")/', $argument, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE) as $part) {
-                if ('"' === $part) {
-                    $escapedArgument .= '\\"';
-                } elseif (self::isSurroundedBy($part, '%')) {
-                    // Avoid environment variable expansion
-                    $escapedArgument .= '^%"'.substr($part, 1, -1).'"^%';
-                } else {
-                    // escape trailing backslash
-                    if ('\\' === substr($part, -1)) {
-                        $part .= '\\';
-                    }
-                    $quote = true;
-                    $escapedArgument .= $part;
-                }
-            }
-            if ($quote) {
-                $escapedArgument = '"'.$escapedArgument.'"';
-            }
+        if (1 === func_num_args()) {
+            $argument = preg_replace_callback("/!([^!=\n]++)!/", function ($m) {
+                @trigger_error(sprintf('Delayed variables are deprecated since Symfony 3.3 and will be left unresolved in 4.0. Resolve the %s variable before calling ProcessUtil::escapeArgument().', $m[0]), E_USER_DEPRECATED);
 
-            return $escapedArgument;
+                return getenv($m[1]);
+            }, $argument);
         }
 
-        return escapeshellarg($argument);
+        if ((self::ESC_WINDOWS_ARGV & $mode) && (false !== strpbrk($argument, " \t\n\v\"") || !isset($argument[0]))) {
```

Comment:
shouldn't the case of the empty string be checked before the ``strpbrk`` ? I suspect it is a cheaper call